test/e2e: Introduce a e2e framework for testing (v2)

wainersm opened the pull request:
**[WORK IN PROGRESS but I'd like to hear anyone comments/suggestions before moving on]**

Cc @bpradipt @jtumber-ibm @stevenhorsman 

This is an implementation in Go of the e2e framework v1 that I proposed in https://github.com/confidential-containers/cloud-api-adaptor/pull/362 .

The v2 implementation is based on the [kubernetes-sigs e2e framework](https://github.com/kubernetes-sigs/e2e-framework) as suggested by @bpradipt . I kept some principles of the v1:

 * use of Go interfaces (instead of abstract class in Python) to abstract pieces that are cloud provider specific (e.g. the code to create a VPC)
 * the e2e is executed with a simple `make test-e2e`

See the test/e2e/README.md added in this pull request for more details about the usage and how the framework should be extended to support other providers.

**Unlike in v1, I did not:**

 * implement checkers for if the local host has the tools (e.g. terraform, kcli,...) required nor the required environment variables were exported.
 * print nice headers to indicate the begin and end of a step. The kubernetes-sigs/e2e-framework can separate nicely between the setup and run tests sections though. This definitively should be improved otherwise users will see a storm of mixed messages
 * implement the build and push of podvm image

**What is new in v2:**

* Implemented  common functions to deploy the operator as suggested by @jtumber-ibm
* Implemented one real test case that can re-used by all cloud providers
* Introduced the idea of "cloud assertion" which are assertions used on the shared tests and which requires interaction with the cloud provider (so they need provider-specific implementations)

**What's missing:**

Well, a bunch of things to be honest, but here are the priorities in my opinion:

 * I'm not sure if this framework should trigger the build of the podvm (see below), but I believe at least it should push the image to the VPC storage
 * Some files in `install/overlays` should be edit with information about the cloud provider, authentication, ssh keys...etc before run `kustomize`. This is very specific to each cloud provider overlay and should be somehow handled by the e2e framework.

**What are the doubts I currently have:**

We can assume the user of the framework (i.e. a developer or CI) will build the entire project (i.e. podvm and operator/runtime) beforehand and have the required tools (e.g. terraform) installed. Or the framework can be doing that, which certainly will make it way more complex.

**Dependencies**

 This PR depends on https://github.com/confidential-containers/cloud-api-adaptor/pull/345

stevenhorsman commented:
Hey Wainer, I've had a quick look through, I think the go e2e framework is going to be a big learning curve to me, but it looks pretty good so far. I have a couple of questions:
- How does the provisioning in MainTest interact with the others? Is it just something in the framework that means this is called before other `*_test.go` files?
- I notice you have the mostly blank `aws_test.go`. Is your thinking that each of us that own a cloud provider could contribute out tests to open-source, but would only be able to run them internal (until we can sort out test cloud accounts), is is the nature of the framework that they'd all get run, so you'd expect stubs that we'd need to fork ourselves?

Thanks for the efforts so far and good luck 🇧🇷

bpradipt commented:
@wainersm finally got a chance to look into this PR. The approach looks good.

huoqifeng commented on test/e2e/provision.go:
```diff
@@ -0,0 +1,90 @@
+
+	overlayDir := path.Join("../../install/overlays", p.cloudProvider)
+	kustomizeHelper := &KustomizeHelper{configDir: overlayDir}
+	if err := kustomizeHelper.Apply(ctx, cfg); err != nil {
```
Regard the overlay when deploy PeerPod, seems it's OK for libvirt provider to use all default values, what about other providers, should user provide the kustomization.yaml and place it in right place before run the test cases?

wainersm replied:
@huoqifeng you are right. Once this PR is merged we could start working on the abstractions for the scenario. Probably we will follow the same pattern: define an interface that each cloud provider should implement.

huoqifeng reviewed (commented):
Thank you @wainersm and @bpradipt for bringing the good approach. I like the idea to use k8s ete framework and give flexibility to setup cluster beforehand or let provisioner to provision it while run the test.

huoqifeng commented:
@wainersm I'll have a try to add ibmcloud_test based on your fork/branch, hopefully this PR can be merged soon :-)

huoqifeng commented:
Added ibmcloud tester in https://github.com/huoqifeng/cloud-api-adaptor/commit/fb2cc4e42ace3a84a94196b0342131750a0eea6c, I'll have a try in IKS cluster.

huoqifeng commented:
I can have the test passed for ibmcloud provider based on changes in https://github.com/huoqifeng/cloud-api-adaptor/commits/ci-e2e-framework-go, Usually it takes 60~120 seconds to finish the case but some rare case took more than 120s. I'm wondering can we increase the time out to 180s? I made change in https://github.com/huoqifeng/cloud-api-adaptor/commit/4c8580ff54ba542c927c1760049864346c6ac515

huoqifeng commented on test/e2e/provision.go:
```diff
@@ -0,0 +1,90 @@
+// CloudProvision defines operations to provision the environment on cloud providers.
+type CloudProvision interface {
+	CreateCluster(ctx context.Context, cfg *envconf.Config) error
+	CreateVPC(ctx context.Context, cfg *envconf.Config) error
```
Looks we need DeleteVPC and DeleteCluster method in this interface.

wainersm replied:
Hi @huoqifeng , added those functions to the interface on my last update to this PR.

huoqifeng commented on test/e2e/main_test.go:
```diff
@@ -0,0 +1,90 @@
+
+	// Run *once* after the tests.
+	testEnv.Finish(func(ctx context.Context, cfg *envconf.Config) (context.Context, error) {
+		// TODO: implement me.
```
Should DeleteVPC and DeleteCluster invoked here when shouldProvisionCluster is true?

wainersm replied:
Also implemented this function with the new DeleteVPC() and DeleteCluster().

huoqifeng commented on test/e2e/provision_libvirt.go:
```diff
@@ -0,0 +1,73 @@
+
+func (l *LibvirtProvisioner) CreateCluster(ctx context.Context, cfg *envconf.Config) error {
+	cmd := exec.Command("/bin/bash", "-c", "./kcli_cluster.sh create")
+	cmd.Dir = "/home/wmoschet/go/src/github.com/confidential-containers/cloud-api-adaptor/libvirt"
```
Is this dir viable?

stevenhorsman replied:
I'm definitely not a go expert, but what about something like:
```
wd, err := os.Getwd()
if err != nil {
      return err
}
cmd.Dir = path.Join(wd, "../../libvirt")
```
?

wainersm replied:
Fixed.

huoqifeng commented:
@bpradipt @wainersm Is there a plan to merge this framework to staging?

huoqifeng commented on test/e2e/provision.go:
```diff
@@ -0,0 +1,90 @@
+		return err
+	}
+
+	if err := AllPodsRunning(ctx, cfg, p.namespace); err != nil {
```
Looks need also make sure RuntimeClasses ready here.

wainersm replied:
hmm... I will fix it.

stevenhorsman commented:
> @bpradipt @wainersm Is there a plan to merge this framework to staging?

@huoqifeng FYI - I think this builds on https://github.com/confidential-containers/cloud-api-adaptor/pull/345 and both that and this PR are ~100 commits behind, so I made a similar status request on that PR. I think Wainer is off until next week, so we might need to wait a little while to get the various bits rebased, updated and then merged. Pradipta might have more info though.

stevenhorsman commented on test/e2e/main_test.go:
```diff
@@ -0,0 +1,90 @@
+	// in the cloud provider. Otherwise, assume the developer did setup the environment and it will
+	// look for a suitable kubeconfig file.
+	shouldProvisionCluster := false
+	if os.Getenv("TEST_E2E_PROVISION") != "" {
```
Not necessarily a complaint, just an observation. I originally tried getting the cluster provisioned and it failed (probably due to user error), so I created it myself and then ran `TEST_E2E_PROVISION=no CLOUD_PROVIDER=libvirt make test-e2e` thinking that `no` would disable provisioning, yet hit a bunch of errors still. Having checked this code I can see that it's not checking for `TEST_E2E_PROVISION=yes`, but for it set at all, so if we keep this behaviour it is probably worth documenting?

wainersm replied:
I will changed it to consider `TEST_E2E_PROVISION=no` case too. thanks!

wainersm replied:
```
diff --git a/test/e2e/main_test.go b/test/e2e/main_test.go
index fc04fd1..5d33253 100644
--- a/test/e2e/main_test.go
+++ b/test/e2e/main_test.go
@@ -34,7 +34,7 @@ func TestMain(m *testing.M) {
        // in the cloud provider. Otherwise, assume the developer did setup the environment and it will
        // look for a suitable kubeconfig file.
        shouldProvisionCluster := false
-       if os.Getenv("TEST_E2E_PROVISION") != "" {
+       if os.Getenv("TEST_E2E_PROVISION") == "yes" {
                shouldProvisionCluster = true
        }

```

wainersm replied:
Fixed.

stevenhorsman commented on test/e2e/main_test.go:
```diff
@@ -0,0 +1,90 @@
+			}
+
+			if err = provisioner.CreateCluster(ctx, cfg); err != nil {
+				return ctx, nil
```
Should this be `return ctx, err` instead as at the moment it's just ignoring the error

wainersm replied:
Fixed.

stevenhorsman commented on test/e2e/provision.go:
```diff
@@ -0,0 +1,90 @@
+	if err != nil {
+		return err
+	}
+	metaList, _ := meta.ExtractList(objList)
```
I *think* the problem might be that this code get two pods initially: `cc-operator-controller-manager-<>` and `cloud-api-adaptor-daemonset-<cloud-provider>-<>`, and waits until they are running, but by then `cc-operator-daemon-install-<>` has been created, but not in running state and we don't go and re-check the list.

wainersm commented:
Updated this PR. Only rebased with staging. The https://github.com/confidential-containers/cloud-api-adaptor/pull/345 is merged so I dropped some commits.

wainersm commented:
Updated this PR to:
 * remove hard-cored path to `/home/wmoschet/go/src/github.com/confidential-containers/cloud-api-adaptor/libvirt`
 * handle `TEST_E2E_PROVISION=no`

wainersm commented:
This update contains:

 * added teardown code (DeleteVPC() and DeleteCluster() added to the CloudProvision interface) as suggested by @huoqifeng . However it is not 100% fine, I noticed when some error occurs on the test or setup then the `testEnv.Finish` block doesn't get called.
 * added a step where the podvm qcow2 can be uploaded to the VPC (commit https://github.com/confidential-containers/cloud-api-adaptor/pull/386/commits/5a337a9a3cfe3779f53430e60aec3ffd4f00f495). At this point the framework won't build the podvm, so assume the user built it.
 * fixed the errors to deploy peer pods, more specifically the code to check it is installed. I added in the commit https://github.com/confidential-containers/cloud-api-adaptor/pull/386/commits/543870785f19219efd58a183e1eaf1e73899dab7 but it can be squashed before the merge

wainersm commented:
@stevenhorsman @huoqifeng the status is that it works for libvirt with some twists (e.g. I had to setup kustomize manually). There still some features that I am working on, e.g., uninstall peer pods from the test environment.

In my biased opinion we could merge this and work on improvements & enablements (e.g. ibmcloud) on follow up PRs.

huoqifeng reviewed (approved):
/lgtm, @wainersm thank you!

bpradipt commented:
@kartikjoshi21

kartikjoshi21 commented:
@surajssd @mkulke

stevenhorsman commented:
> * fixed the errors to deploy peer pods, more specifically the code to check it is installed. I added in the commit [5438707](https://github.com/confidential-containers/cloud-api-adaptor/commit/543870785f19219efd58a183e1eaf1e73899dab7) but it can be squashed before the merge

Hi Wainer, I tried it out and still hit:
```
Wait at least one worker be Ready
peer-pods-worker-0     Ready    worker          51s     v1.26.1
Wait system pods be running or completed
Deploy Peer Pods
Install the controller manager
Wait for the cc-operator-controller-manager deployment be available
Install CoCo and cloud-api-adaptor
Wait for at least one pod of the cc-operator-daemon-install daemonset be ready
Wait for at least one pod of the cloud-api-adaptor-daemonset-libvirt daemonset be ready
Check the kata runtimeclass exists
F0208 14:19:26.070964   80523 env.go:360] Setup failure: runtimeclasses.node.k8s.io "kata" not found
goroutine 1 [running]:
```

Any ideas if I might be doing something wrong?

wainersm commented:
> > * fixed the errors to deploy peer pods, more specifically the code to check it is installed. I added in the commit [5438707](https://github.com/confidential-containers/cloud-api-adaptor/commit/543870785f19219efd58a183e1eaf1e73899dab7) but it can be squashed before the merge
> 
> Hi Wainer, I tried it out and still hit:
> 
> ```
> Wait at least one worker be Ready
> peer-pods-worker-0     Ready    worker          51s     v1.26.1
> Wait system pods be running or completed
> Deploy Peer Pods
> Install the controller manager
> Wait for the cc-operator-controller-manager deployment be available
> Install CoCo and cloud-api-adaptor
> Wait for at least one pod of the cc-operator-daemon-install daemonset be ready
> Wait for at least one pod of the cloud-api-adaptor-daemonset-libvirt daemonset be ready
> Check the kata runtimeclass exists
> F0208 14:19:26.070964   80523 env.go:360] Setup failure: runtimeclasses.node.k8s.io "kata" not found
> goroutine 1 [running]:
> ```
> 
> **Update** The reason is:
> 
> ```
> # kubectl get pods -n confidential-containers-system
> NAME                                             READY   STATUS             RESTARTS        AGE
> cc-operator-controller-manager-fbb5dcf9d-pr57d   2/2     Running            1 (10m ago)     11m
> cc-operator-daemon-install-w4xr6                 1/1     Running            0               10m
> cloud-api-adaptor-daemonset-libvirt-jnzfp        0/1     CrashLoopBackOff   6 (3m38s ago)   10m
> ```
> 
> caused by:
> 
> ```
> # kubectl logs cloud-api-adaptor-daemonset-libvirt-jnzfp -n confidential-containers-system
> + exec cloud-api-adaptor-libvirt libvirt -uri 'qemu+ssh://root@192.168.122.1/system?no_verify=1' -data-dir /opt/data-dir -pods-dir /run/peerpod/pods -network-name default -pool-name default -socket /run/peerpod/hypervisor.sock
> 2023/02/08 14:25:25 [helper/hypervisor] hypervisor config {/run/peerpod/hypervisor.sock   /run/peerpod/pods libvirt}
> 2023/02/08 14:25:25 [helper/hypervisor] cloud config {qemu+ssh://root@192.168.122.1/system?no_verify=1 default default /opt/data-dir}
> 2023/02/08 14:25:25 [helper/hypervisor] Unable to create libvirt connection: virError(Code=38, Domain=7, Message='Cannot recv data: Failed to add the host to the list of known hosts (/root/.ssh/known_hosts).
> root@192.168.122.1: Permission denied (publickey).: Connection reset by peer')
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5244b7]
> 
> goroutine 37 [running]:
> github.com/confidential-containers/cloud-api-adaptor/cmd.(*starter).Start.func2()
> 	/go/cloud-api-adaptor/cmd/starter.go:78 +0x77
> created by github.com/confidential-containers/cloud-api-adaptor/cmd.(*starter).Start
> 	/go/cloud-api-adaptor/cmd/starter.go:75 +0x1a7
> ```
> 
> Any ideas what I did wrong?

The framework still doesn't setup all the kustomize files so you need to do some changes manually. It is failing in your environment because CAA cannot connect to libvirt via SSH (`qemu+ssh://root@192.168.122.1/system?no_verify=1`).

Try the following and you will notice it fails:
```
$ virsh -c "qemu+ssh://root@192.168.122.1/system?no_verify=1" nodeinfo
```

You will need to:
 * Set the ssh pub key as explained [here](https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/libvirt/README.md#install-and-configure-confidential-containers-and-cloud-api-adaptor-in-the-cluster)
 * Uncomment the line https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/install/overlays/libvirt/kustomization.yaml#L32 (just realized that instruction is missing on the libvirt's howto doc)

Anyway, the installation checker is still not fully correct. It should had failed on `Wait for at least one pod of the cloud-api-adaptor-daemonset-libvirt daemonset be ready` , luckily the runtimeclass checker caught the error. I will try something to fix that.

stevenhorsman commented:
> Uncomment the line https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/install/overlays/libvirt/kustomization.yaml#L32 (just realized that instruction is missing on the libvirt's howto doc)

This was the crucial missing step I think - when I ran the process last time I use ./install_operator to set provision the cluster first to test that, so I guess that makes that change?

stevenhorsman reviewed (approved):
This is working for me now. To test it out I did the following on a `bx2-8x32` VSI:
```
export GOPATH="${HOME}/go"
repo="github.com/wainersm/cc-cloud-api-adaptor"
repo_dir="${GOPATH}/src/${repo}"
git clone -b ci-e2e-framework-go "https://${repo}.git" ${repo_dir}
echo "Install podman"
echo "deb https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_20.04/ /" | sudo tee /etc/apt/sources.list.d/devel:kubic:libcontainers:stable.list
curl -L "https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_20.04/Release.key" | sudo apt-key add -
sudo apt update
sudo apt install podman
podman --version
alias docker=podman
echo "Build ppvm image"
pushd ${repo_dir}/podvm
docker build --no-cache -t podvm_builder -f Dockerfile.podvm_builder .
docker build --no-cache --device=/dev/kvm -t podvm_libvirt --build-arg BUILDER_IMG=podvm_builder:latest --build-arg CLOUD_PROVIDER=libvirt -f Dockerfile.podvm .
podvm_container=$(docker create podvm_libvirt:latest /bin/bash)
file_name=$(docker export ${podvm_container} | tar -t | grep podvm)
docker cp ${podvm_container}:${file_name} podvm.qcow2
docker rm ${podvm_container}
export IMAGE=$PWD/podvm.qcow2
ls -al $IMAGE
echo "Install make, go, libvirt, kcli, other packages"
sudo apt install qemu-kvm libvirt-daemon-system libvirt-clients bridge-utils virtinst virt-manager
sudo usermod -aG libvirt $USER
sudo usermod -aG kvm $USER
curl https://raw.githubusercontent.com/karmab/kcli/main/install.sh | sudo bash
sudo kcli create pool -p /var/lib/libvirt/images default
sudo setfacl -m u:$(id -un):rwx /var/lib/libvirt/images
virsh vol-create-as --pool default --name podvm-base.qcow2 --capacity 20G --allocation 2G --prealloc-metadata --format qcow2
virsh vol-upload --vol podvm-base.qcow2 $IMAGE --pool default --sparse
virsh -c qemu:///system vol-info --pool default podvm-base.qcow2
sudo apt-get install -y libvirt-dev
sudo apt-get install -y build-essential
curl -OL https://go.dev/dl/go1.19.5.linux-amd64.tar.gz
rm -rf /usr/local/go && tar -C /usr/local -xzf go1.19.5.linux-amd64.tar.gz
export PATH=$PATH:/usr/local/go/bin
go version
curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
cd ../install/overlays/libvirt
ssh-keygen -f ./id_rsa -N ""
cat id_rsa.pub >> ~/.ssh/authorized_keys
chmod 600 ~/.ssh/authorized_keys
virsh -c "qemu+ssh://$USER@192.168.122.1/system?keyfile=$(pwd)/id_rsa" nodeinfo
echo "Uncomment line    #- id_rsa # set - path to private key of kustomization.yaml"
vim kustomization.yaml
cd ../../../
ssh-keygen -t rsa -C "stevenh-libvirt-pp"
TEST_E2E_PROVISION=yes CLOUD_PROVIDER=libvirt make test-e2e
```

The result of this was:
```
Wait at least one worker be Ready
peer-pods-worker-0     Ready    worker          54s     v1.26.1
Wait system pods be running or completed
Deploy Peer Pods
Install the controller manager
Wait for the cc-operator-controller-manager deployment be available
Install CoCo and cloud-api-adaptor
Wait for at least one pod of the cc-operator-daemon-install daemonset be ready
Wait for at least one pod of the cloud-api-adaptor-daemonset-libvirt daemonset be ready
Check the kata runtimeclass exists
=== RUN   TestLibvirtCreateSimplePod
=== RUN   TestLibvirtCreateSimplePod/Simple_Peer_Pod
=== RUN   TestLibvirtCreateSimplePod/Simple_Peer_Pod/PodVM_is_created
--- PASS: TestLibvirtCreateSimplePod (50.04s)
    --- PASS: TestLibvirtCreateSimplePod/Simple_Peer_Pod (50.04s)
        --- PASS: TestLibvirtCreateSimplePod/Simple_Peer_Pod/PodVM_is_created (0.00s)
PASS
Deleting directory /root/.kcli/clusters/peer-pods
peer-pods-ctlplane-0 deleted on local!
peer-pods-worker-0 deleted on local!
ok  	github.com/confidential-containers/cloud-api-adaptor/test/e2e	568.793s
```
Thanks for the efforts @wainersm !

wainersm commented:
> > Uncomment the line https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/install/overlays/libvirt/kustomization.yaml#L32 (just realized that instruction is missing on the libvirt's howto doc)
> 
> This was the crucial missing step I think - when I ran the process last time I use ./install_operator to set provision the cluster first to test that, so I guess that makes that change?

yep, `./install_operator.sh` changed that file. Probably you reseted your tree.

wainersm commented:
ok, on the last updates I fixed the fix for the installation checkers. Now the code wait for the daemonset to schedule at least one pod, then wait each pod to be ready.

I ran it locally, it worked out:
```
peer-pods-master-0 deployed on local
Deploying workers
Deploying Vms...
peer-pods-worker-0 deployed on local
Kubernetes cluster peer-pods deployed!!!
INFO export KUBECONFIG=$HOME/.kcli/clusters/peer-pods/auth/kubeconfig
INFO export PATH=$PWD:$PATH
Wait at least one worker be Ready
peer-pods-worker-0   Ready    worker          38s     v1.26.1
Wait system pods be running or completed
Deploy Peer Pods
Install the controller manager
Wait for the cc-operator-controller-manager deployment be available
Install CoCo and cloud-api-adaptor
Wait for the pod cc-operator-daemon-install-bscwb be ready
Wait for the pod cloud-api-adaptor-daemonset-libvirt-cxbd8 be ready
Check the kata runtimeclass exists
=== RUN   TestLibvirtCreateSimplePod
=== RUN   TestLibvirtCreateSimplePod/Simple_Peer_Pod
=== RUN   TestLibvirtCreateSimplePod/Simple_Peer_Pod/PodVM_is_created
--- PASS: TestLibvirtCreateSimplePod (80.09s)
    --- PASS: TestLibvirtCreateSimplePod/Simple_Peer_Pod (80.09s)
        --- PASS: TestLibvirtCreateSimplePod/Simple_Peer_Pod/PodVM_is_created (0.04s)
PASS
Deleting directory /home/wmoschet/.kcli/clusters/peer-pods
peer-pods-master-0 deleted on local!
peer-pods-worker-0 deleted on local!
ok      github.com/confidential-containers/cloud-api-adaptor/test/e2e   851.665s
```

I assume that @stevenhorsman and @huoqifeng 's approvals stand, so gonna merge it. I will send a follow up PR soon to include features left behind, so if we find additional bugs in meanwhile then I can address them on that PR.
